stan/math/prim/fun/multiply.hpp

```diff
@@ -70,6 +70,30 @@ inline auto multiply(const Mat1& m1, const Mat2& m2) {
   return (m1 * m2).eval();
 }
 
+/**
+ * Return the product of the specified matrices. The number of
+ * columns in the first matrix must be the same as the number of rows
+ * in the second matrix.
+ *
+ * @tparam Mat1 type of the first matrix or expression
+ * @tparam Mat2 type of the second matrix or expression
+ *
+ * @param m1 first matrix or expression
+ * @param m2 second matrix or expression
+ * @return the product of the first and second matrices
+ * @throw <code>std::invalid_argument</code> if the number of columns of m1 does
+ * not match the number of rows of m2.
+ */
+template <typename Mat1, typename Mat2,
+          require_all_eigen_t<Mat1, Mat2>* = nullptr,
+          require_t<is_complex<return_type_t<Mat1, Mat2>>>* = nullptr,
+          require_not_eigen_row_and_col_t<Mat1, Mat2>* = nullptr>
+inline auto multiply(const Mat1& m1, const Mat2& m2) {
+  check_size_match("multiply", "Columns of m1", m1.cols(), "Rows of m2",
+                   m2.rows());
+  return m1.lazyProduct(m2).eval();
```

Comment:
To use regular product here we would need to specialize `general_matrix_matrix_product` and `general_matrix_vector_product` in `Eigen_NumTraits.hpp`. `lazyProduct()` is inefficient but handles promotion nicer so that complex matrices and regular matrices work. tbh I just didn't want to write the gemm specializations because they are complicated but in the future we can do that